### PR TITLE
Allow to not optimize release build

### LINKED ISSUE #2457: Add fast compilation option for target=`release`
### Describe the project you are working on
Godot Engine
### Describe the problem or limitation you are having in your project
Compilation with `target=release` takes a lot of time, even on multicore processor.
### Describe the feature / enhancement and how it helps to overcome the problem or limitation
In Godot should be available option to disable optimizations for testing purposes.
Only `target=release` doesn't add `DEBUG_ENABLED` Godot flag, so isn't possible to test how app behave after exporting using different options.
### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Probably `optimize` scons variable should allow to set `none` to it.
### If this enhancement will not be used often, can it be worked around with a few lines of script?
It is easy to workaround this, but it will be better to have an official switch
### Is there a reason why this should be core and not an add-on in the asset library?
This cannot be added as addon.

### PR DESCRIPTION
**Pull request by qarmin**
Fixes https://github.com/godotengine/godot-proposals/issues/2457

This will allow to speedup compilation for people, that want to test Godot without `DEBUG_ENABLED` flag

If fix is good, then I will do the same for others platform

### CODE REVIEW

File: platform/linuxbsd/detect.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def configure(env):
     if env["target"] == "release":
         if env["optimize"] == "speed":  # optimize for speed (default)
             env.Prepend(CCFLAGS=["-O3"])
-        else:  # optimize for size
+        elif env["optimize"] != "none":  # optimize for size
```
Comment (Xrayez) [inline platform/linuxbsd/detect.py:93]:
```suggestion
        elif env["optimize"] == "size":  # optimize for size
```
No optimization flags are going to be applied with `none` as it won't be handled.

File: platform/linuxbsd/detect.py [hunk 2]
```diff
@@ -99,7 +99,7 @@ def configure(env):
     elif env["target"] == "release_debug":
         if env["optimize"] == "speed":  # optimize for speed (default)
             env.Prepend(CCFLAGS=["-O2"])
-        else:  # optimize for size
+        elif env["optimize"] != "none":  # optimize for size
```
Comment (Xrayez) [inline platform/linuxbsd/detect.py:102]: Same as above.

```suggestion
        elif env["optimize"] == "size":  # optimize for size
```

**Comment by akien-mga** [comment]
Seems fine to me (with the changes suggested by @Xrayez), so you can apply it to other platforms.

**Comment by qarmin** [comment]
Compilation flags between 3.2 and 4.0 branch are very similar so I added cherrypick(probably only linuxbsd->x11 needs to be done).

I tested in 3.2 branch compilation times with default optimization and this took more than 4h in CI - https://github.com/qarmin/RegressionTestProject/pull/29/checks?check_run_id=2102453232
Probably after adding this commit, compilation times should go down to ~30min-1h

File: platform/android/detect.py [hunk 3]
```diff
@@ -198,7 +198,7 @@ def mySpawn(sh, escape, cmd, args, env):
             env.Append(LINKFLAGS=["-O2"])
             env.Append(CCFLAGS=["-O2", "-fomit-frame-pointer"])
             env.Append(CPPDEFINES=["NDEBUG"])
-        else:  # optimize for size
+        elif env["optimize"] == "size":  # optimize for size
             env.Append(CCFLAGS=["-Os"])
             env.Append(CPPDEFINES=["NDEBUG"])
```
Comment (akien-mga) [inline platform/android/detect.py:203]: The `NDEBUG` define should be moved out of the `optimize` blocks IMO.

File: platform/javascript/detect.py [hunk 4]
```diff
@@ -71,8 +71,9 @@ def configure(env):
         # -Os reduces file size by around 5 MiB over -O3. -Oz only saves about
         # 100 KiB over -Os, which does not justify the negative impact on
         # run-time performance.
-        env.Append(CCFLAGS=["-Os"])
-        env.Append(LINKFLAGS=["-Os"])
+        if env["optimize"] != "none":
```
Comment (akien-mga) [inline platform/javascript/detect.py:74]: Should also be added to `release_debug` branch. Might be worth refactoring into `if env["target"].startswith("release"):` like done for iphone.

File: platform/linuxbsd/detect.py (hunk 2 above)
Comment (akien-mga) [reply platform/linuxbsd/detect.py:102]: This wasn't changed.

File: platform/osx/detect.py [hunk 5]
```diff
@@ -50,7 +50,7 @@ def configure(env):
     if env["target"] == "release":
         if env["optimize"] == "speed":  # optimize for speed (default)
             env.Prepend(CCFLAGS=["-O3", "-fomit-frame-pointer", "-ftree-vectorize"])
-        else:  # optimize for size
+        elif env["optimize"] == "size":  # optimize for size
```
Comment (akien-mga) [inline platform/osx/detect.py:53]: Should be done for `release_debug` too.

File: platform/server/detect.py [hunk 6]
```diff
@@ -56,7 +56,7 @@ def configure(env):
     if env["target"] == "release":
         if env["optimize"] == "speed":  # optimize for speed (default)
             env.Prepend(CCFLAGS=["-O3"])
-        else:  # optimize for size
+        elif env["optimize"] == "size":  # optimize for size
```
Comment (akien-mga) [inline platform/server/detect.py:59]: `release_debug` too.

File: platform/uwp/detect.py [hunk 7]
```diff
@@ -54,9 +54,11 @@ def configure(env):
     ## Build type
 
     if env["target"] == "release":
-        env.Append(CCFLAGS=["/O2", "/GL"])
         env.Append(CCFLAGS=["/MD"])
-        env.Append(LINKFLAGS=["/SUBSYSTEM:WINDOWS", "/LTCG"])
+        env.Append(LINKFLAGS=["/SUBSYSTEM:WINDOWS"])
+        if env["optimize"] != "none":
```
Comment (akien-mga) [inline platform/uwp/detect.py:59]: `release_debug` too.

File: platform/windows/detect.py [hunk 8]
```diff
@@ -190,10 +190,11 @@ def configure_msvc(env, manual_msvc_config):
     if env["target"] == "release":
         if env["optimize"] == "speed":  # optimize for speed (default)
             env.Append(CCFLAGS=["/O2"])
-        else:  # optimize for size
+            env.Append(LINKFLAGS=["/OPT:REF"])
+        elif env["optimize"] == "size":  # optimize for size
```
Comment (akien-mga) [inline platform/windows/detect.py:194]: `release_debug` too.

**Review by Calinou — APPROVED** [review]
In the future, we could extend this to editor builds to get fast compilation feedback (at the cost of very slow runtime performance).

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.3.

**Pull request merged**